### DISCUSSION TITLE
Audio Data Java/Kotlin C++ Interaction Questions

### DISCUSSION THREAD

**Discussion by lilinxiong** [google/oboe#1375] [Q&A] [answered]
Thank you very much to the Oboe team for opening up such a great project. I now have some questions that I need to ask.
I have read the [Can I write audio data from Java/Kotlin to Oboe?](https://github.com/google/oboe/blob/master/docs/FAQ.md#can-i-write-audio-data-from-javakotlin-to-oboe) I tentatively asked, is it not recommended Audio Data from java/kotlin to Oboe? The same is not recommended Audio Data from Oboe to java/kotlin? 
In other words, they can't work? Because I see that an ``onAudioReady`` callback is needed in Oboe. Oboe itself is a complete closed loop. Whether data to ``java/kotlin`` or data to ``Oboe`` will destroy this loop, it will not ``get a low latency stream``.
If you want to use Oboe, then all audio processing needs to use C++. Only in this way, you will eventually get a low latency stream.
Is my understanding correct?
In [Can I write audio data from Java/Kotlin to Oboe?](https://github.com/google/oboe/blob/master/docs/FAQ.md#can-i-write-audio-data-from-javakotlin-to-oboe) I saw if I want to get a low latency stream in the java layer, then I can use the AudioTrack.Builder method ``setPerformanceMode(AudioTrack.PERFORMANCE_MODE_LOW_LATENCY)``.
I would like to ask, if I use the ``AudioTrack`` method in java, will I get the same stream as the ``Oboe`` low latency stream? If the delay of ``AudioTrack`` in java is higher than that of ``Oboe``, how much higher is it?
We want to do ear return, but now there are many methods in the java layer, it is difficult to transplant in a short time.
thank you very much. Looking forward to the response from the Oboe team.

**Comment by philburk** [comment, marked as answer]
It is possible to send data in either direction between Java and Oboe.
If you use a callback then you will probably want to use a FIFO between Java and the Oboe callback. Atomic FIFOs can be implemented using a ByteBuffer.
If you read or write to Oboe then you do that directly in a JNI call from Java.

> I would like to ask, if I use the AudioTrack method in java, will I get the same stream as the Oboe low latency stream?

Not exactly. If you are writing to the stream and not using a callback then the Java path might have lower latency.
Oboe will not provide a LowLatency output stream for write() unless the device supports MMAP. On the Legacy path you will get a normal stream with higher latency than Java can get using AudioTrack.PERFORMANCE_MODE_LOW_LATENCY.

Rough comparison of output latency (varies by device)
Oboe LowLatency with callback using MMAP = 10-18 msec
Oboe LowLatency with callback using Legacy = 14-50 msec
Java AudioTrack.PERFORMANCE_MODE_LOW_LATENCY = 40-100 msec
Oboe using write on Legacy path = 100-300 msec

**Reply by lilinxiong** [reply]
Thank you for your answer, I got what I want, thank you again~

**Reply by jet-yangqs** [reply]
if I use AudioTrack.PERFORMANCE_MODE_LOW_LATENCY in java on LowLatency path, can I use the Audio effect such as AEC?

**Comment by huangxin1985** [comment]
so...we have to synthesis audio in the jni layer. right?
